### ISSUE TITLE
Problems building Theia under Windows 10

### ISSUE THREAD

**Issue by UdoW** [theia-ide/theia#188]
I tried to build Theia under Window 10 with following setup:

- Node 8.0.0
- npm 4.5.0
- tsc 2.3.2
- GIT Bash : 2.10.0.windows.1

I pulled the newest version of Theia some minutes ago (2017-06-12).

I followed the steps in doc/Developing.md:

Everything worked fine until trying the examples:

First I tried example Browser:
Step npm run bootstrap worked fine.
Step npm start ended in missing dependencies
 

- concurrently
- webpack-dev-server
- bunyan
- webpack
- rimraf

And after installing these modules an error message was printed that the lib folder ist missing.

I tried then a npm run cold:start and got following error:

browser-example@0.0.1-alpha.1 lint D:\git_2\external\theia\examples\browser
tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json
 
Warning: The 'no-use-before-declare' rule requires type checking
 
ERROR: src/backend/main.ts[41, 1]: unused expression, expected an assignment or function call
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! browser-example@0.0.1-alpha.1 lint: `tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the browser-example@0.0.1-alpha.1 lint script 'tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the browser-example package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json

I have no idea how to continue.

Then I tried the electron example: 
I had to install the electron module.

After running npm start I got a Windows error dialog with the message:
Unable to find Electron app at
D:\git_2\external\theia\examples\electron\lib\frontend\electron-main.js

Then I tried also npm run cold:start and got the same message as for the browser example:

$ npm run cold:start
 
> electron-example@0.0.1-alpha.1 cold:start D:\git_2\external\theia\examples\electron
> npm run clean && npm run build && npm start
 
 
> electron-example@0.0.1-alpha.1 clean D:\git_2\external\theia\examples\electron
> rimraf lib
 
 
> electron-example@0.0.1-alpha.1 build D:\git_2\external\theia\examples\electron
> npm run build:localdeps && npm run build:backend && npm run build:frontend
 
 
> electron-example@0.0.1-alpha.1 build:localdeps D:\git_2\external\theia\examples\electron
> ldm run build && ldm sync
 
theia-core: npm run build
 
> electron-example@0.0.1-alpha.1 build:backend D:\git_2\external\theia\examples\electron
> tsc && npm run lint
 
 
> electron-example@0.0.1-alpha.1 lint D:\git_2\external\theia\examples\electron
> tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json
 
Warning: The 'no-use-before-declare' rule requires type checking
 
ERROR: src/backend/main.ts[41, 1]: unused expression, expected an assignment or function call
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! electron-example@0.0.1-alpha.1 lint: `tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the electron-example@0.0.1-alpha.1 lint script 'tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the electron-example package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs electron-example
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls electron-example
npm ERR! There is likely additional logging output above.

How can I build Theia under Windows 10? Do I have to install different versions or what is wrong in my setup?

With regards,
Udo

**Comment by kittaakos** [comment]
@UdoW,  first off, thanks for trying it out!

Currently, one [needs npm4] to be able to install and run the examples. I can see that you are using npm4 and node8. Please correct me if I am wrong, but node8 goes with npm5 by default, right? Are you sure you have npm4? Could you please try to run it with node7 and npm4? I can confirm, that combination worked fine.

Meanwhile, I try it out on a Win10 virtual box.

[needs npm4]: https://github.com/theia-ide/theia/issues/179

**Comment by UdoW** [comment]
How can I restrict it to Node7?

**Comment by kittaakos** [comment]
I am just trying to reproduce your problem on Win10. I'll update this thread soon. Bear with me.

**Comment by kittaakos** [comment]
@UdoW, I can confirm, it fails for me on Win10 virtual box too. I have a completely different issue. I used Node 7.9.0 and npm 4.2.0.

Could you please share your entire command line output with us as a `txt` file? If so, just attach it to this thread. Thanks in advance for your cooperation!

**Comment by UdoW** [comment]
Logging file from ´npm run cold:start´ in folder examples/electron

[2017-06-12T08_52_44_788Z-debug.txt](https://github.com/theia-ide/theia/files/1067562/2017-06-12T08_52_44_788Z-debug.txt)

**Comment by UdoW** [comment]
I just upgraded to npm 5.0.3. Same error message.

**Comment by kittaakos** [comment]
By attaching the entire log; I meant something like [this](https://github.com/theia-ide/theia/files/1067597/gh_188_theia_install_win10_virtual_box.txt).

**Comment by UdoW** [comment]
Then I downgraded to npm 4.2.0. Now everything crashes, I mean also the npm command crashes in itself.... :(

I first have to fix my installation. Then I'll send you the full log file.

**Comment by kittaakos** [comment]
Cool, thanks! We definitely have issues on Windows, no questions at all.
It seems, `concurrently` is not installed with the local dependency manager on Windows.
CCing @akosyakov

**Comment by UdoW** [comment]
Running browser example.

[example_browser_console_log.txt](https://github.com/theia-ide/theia/files/1067663/example_browser_console_log.txt)

Then trying to run "npm run cold:start"

[example_browser_cold_start_console_log.txt](https://github.com/theia-ide/theia/files/1067682/example_browser_cold_start_console_log.txt)

**Comment by kittaakos** [comment]
I have managed to run it on Win10 but our steps are clearly broken for Windows. I will write down the steps in short. As I see you had the same problem as I did.

**Comment by kittaakos** [comment]
- Make sure you have node v7.x.x with npm 4.x.x on you machine. Download node v7.9.0 for Windows from [here](https://nodejs.org/dist/v7.9.0/node-v7.9.0-x64.msi).
 - Verify the versions: `node -v && npm -v`. If you see for instance `v7.9.0 4.2.0`, then you are fine.
 - Also, make sure you have the required C++ compiler and Python 2.7 for building the native node modules. Run `npm install --global windows-build-tools` as an administrator on your system.
 - Then run these steps. Please note, it differs from the steps inidcated in the `tl;dr` scripts:

```
   git clone https://github.com/theia-ide/theia \
   && cd theia \
   && npm install \
   && cd config/local-dependency-manager \
   && npm install \
   && cd ../../examples/browser \
   && npm install \
   && npm run bootstrap \
   && npm run start
```

**Comment by UdoW** [comment]
I install the versions you gave me.

I had to install the modules undertaker and typescript.

Now the build "npm install" in the examples/browser folder hangs. It does not find "ldm". 
I installed it with "npm -g install ldm" but it still cannot be found.

**Comment by kittaakos** [comment]
> I had to install the modules undertaker and typescript.

Why was it necessary?

> I installed it with "npm -g install ldm" but it still cannot be found.

Instead of installing `ldm` to the global, could you just execute the steps from above? It worked for me.

**Comment by UdoW** [comment]
Hmm, for me the steps did not work.  The 3rd "npm install" did not finish successfully.

**Comment by UdoW** [comment]
tsc is still missing.

**Comment by kittaakos** [comment]
> The 3rd "npm install" did not finish successfully.

Does it hang or does it complain about `tsc`? What exactly the problem in the third `npm i` step? By the way, that additional step fixed the problem for me.

Do you have `tsc` in the global? It is not required and could be the cause of the problem. (But I am still just guessing...)

**Comment by UdoW** [comment]
Hmm, I installed node 8.1.0 again and tried again and I still have the same log output:

> browser-example@0.0.1-alpha.1 clean D:\theia_n810_npm503\theia\examples\browser
> rimraf lib


> browser-example@0.0.1-alpha.1 build D:\theia_n810_npm503\theia\examples\browser
> npm run build:localdeps && npm run build:backend && npm run build:frontend


> browser-example@0.0.1-alpha.1 build:localdeps D:\theia_n810_npm503\theia\examples\browser
> ldm run build && ldm sync

Der Befehl "ldm" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! browser-example@0.0.1-alpha.1 build:localdeps: `ldm run build && ldm sync`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the browser-example@0.0.1-alpha.1 build:localdeps script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\uwalker\AppData\Roaming\npm-cache\_logs\2017-06-12T13_35_12_582Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! browser-example@0.0.1-alpha.1 build: `npm run build:localdeps && npm run build:backend && npm run build:frontend`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the browser-example@0.0.1-alpha.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

**Comment by UdoW** [comment]
I will try it tomorrow again.

**Comment by UdoW** [comment]
I tried it again with node 7.9.0 and npm 4.2.0 from your installer:

I still get the same error as when I use version 8.1.0 and 5.0.3. 

Which shell did you use? I tried it in MINGW64 and Cygwin64 Terminal. In MINGW64it hangs at some point and in Cygwin64 it gives the same error as with version 8.1.0/5.0.3.

**Comment by kittaakos** [comment]
> Which shell did you use?

Upfront, I have to verify I am not a Windows power user. I used the GitBash integrated into the `cmd.exe`. Besides it, I have installed exactly the same node installer I have referenced somewhere above; so I have node `7.9.0` and npm `4.2.0` on my path. And I run the [TL;DR](https://github.com/theia-ide/theia/blob/master/doc/Developing.md#tldr) script.

**Comment by kittaakos** [comment]
By the way, we have identified and fixed an issue with `ldm`. See details here: #193.

**Comment by UdoW** [comment]
I am still stuck on the "npm install" of the browser examples: It cannot find ldm.

How do I have to install ldm so it can be found on my machine?

**Comment by kittaakos** [comment]
I think you do not need to explicitly install it. At least I've never done that before and it still works for me both on Windows and OS X. @akosyakov, could you please give advice?

**Comment by akosyakov** [comment]
@UdoW Have you run `npm run bootstrap` from the browser example?

If you want `ldm` globally:
- go to `config/local-dependency-manager`
- `npm install` to install ldm dependencies
- `npm install -g .` to install ldm globally

**Comment by kittaakos** [comment]
But it is not necessary to have `ldm` globally, right? It should work without it too. At least that was my experience so far.

**Comment by akosyakov** [comment]
right, if you do `npm run bootstrap`

**Comment by kittaakos** [comment]
@UdoW, please do the followings:  perform a `git pull` so you have the latest changes (with the fix for #193). Then execute the tldr script. If it still fails for you, please share with the full log from the terminal. Also, please do a `node -v && npm -v` before you execute the script, so we can have all the information.

**Comment by kittaakos** [comment]
Another thing came to my mind; we had problems with `ldm` on Windows. So make sure you uninstall it from the global. You can re-install it later if you need that.

**Comment by kittaakos** [comment]
> I got it somehow running but I have to replay again to tell you the steps.

Excellent. We need to verify whether it works with the given steps or not.

For the rest with the terminal; please open a separete issue and I look into that.

**Comment by kittaakos** [comment]
@UdoW, did you manage to reproduce your build problem or it is already solved with #193? Thanks!

**Comment by kittaakos** [comment]
I close this issue. If one still has problem installing, building or running Theia, please reopen.
